### ISSUE TITLE
Content-id derivation function

### ISSUE THREAD

**Issue by victor-wei126** [ethereum/trin#98]
So in order to get the content-id from the content-key, there is some function that maps the key to the id. On the history network, this is just simply taking the sha256 hash of the entire key. On the state network, it is more complicated since depending on the type of content, we may use different hash functions, which may hash different portions of the content-key.

So I am thinking this method will be defined in `utils.rs` and it will take in a key as parameter, and return the id after performing appropriate computation. To do the right computation, it needs to determine 1) if this key is for content from history or state and 2) if from state, what type of content exactly is it (account trie node, contract trie node). 

For 2), the function can just extract the first byte from the key, since that is the `content_type` and be done with it there. For 1, I am not so sure how to do this. If the only parameter is the key, how can we differentiate between the networks?

**Comment by njgheorghita** [comment]
If I understand correctly, I don't see why we couldn't have separate derivation utils within their respective package (`trin-state`, `trin-history`). So, deriving content-ids will be encapsulated within a subnetworks logic (i.e. `derive_content_id()` within `trin-state/utils.rs` will only have to worry about deriving state network ids, etc.).

**Comment by victor-wei126** [comment]
I think that makes a lot of sense. Although, I thought about this some more and I believe @mrferris implementation of should_store() method will call these derivation functions, and so depending on whether the key came from a state network or history network message will still have to be determined either way. Unless, this is just purely an issue of keeping things encapsulated, then that makes sense.

**Comment by mrferris** [comment]
Right now I'm working on making the PortalStorage's content-id function parameter-izable, so (if we end up going the route of one PortalStorage instance per network) it will look something like:
```
fn history_network_content_key_to_id(&String key) -> String {
...
}
history_storage_config.content_key_to_id = &network_specific_content_key_to_id;
```

And then should_store() and all of the other functions that also need to to do the conversion just call that function internally.

I'm thinking of doing it this way because I don't think that we should be importing from any of the individual network packages into `trin-core` (where `storage.rs` is), and should instead only import `trin-core` into the individual network packages (and `trin` itself). Let me know if anybody understands differently.

If we end up having one instance of PortalStorage handle the data for all networks, then yes @victor-wei126 I was thinking the same: we'll need another way to determine which network the data is for each time we call should_store etc.

**Comment by victor-wei126** [comment]
> I'm thinking of doing it this way because I don't think that we should be importing from any of the individual network packages into trin-core (where storage.rs is), and should instead only import trin-core into the individual network packages (and trin itself). Let me know if anybody understands differently.

Yeah that makes sense to me. Since if we have 2 PortalStorage structs, one for each network, then we can define the derivation functions in `trin-core` without having the state differentiation logic. The only case for still defining the functions in their respective `trin-state` and `trin-history` packages is if some other code in those packages needs to call them. But I can't think of any use cases, and either way we can easily move them to those packages in the future if need be.

**Comment by mrferris** [comment]
I still think that we should define each derivation function in its corresponding package, not in `trin-core`. 

My thinking is basically:
If we aren't going to store things that are specific to the history network in `trin-history`, then what are we going to put in `trin-history`? 

`trin-core` can import them and just pass them into the PortalStorage structs like you said, but the code will read much more clearly if it just passes in a call to a single function instead of defining it there. It also makes it so that we don't have to modify `trin-core` when we're really just modifying the behavior of a single network's derivation function.
